Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -2007,26 +2007,33 @@
 
     imgRenderTarget->BeginDraw();
 
-    // Apply the required transformations as set in the context.
-    // We need some special handling in transform as CoreText in iOS renders from bottom left but DWrite on Windows does top left.
-    //     1. Scaling - to handle scaling once text has been scaled, apply translation to the scaled height
-    //     2. Rotation - CoreText rotates text anti-clockwise but DWrite performs clockwise rotation
-    CGAffineTransform transform = CGAffineTransformConcat(curState->curTransform, curState->curTextMatrix);
-    // Perform translation required to handle scaling.
-    CGFloat verticalScalingFactor = sqrt((transform.b * transform.b) + (transform.d * transform.d));
-    // TODO::
-    // Issue #1057 - CG today talks in Pixels but DirectWrite and D2D only understand DIPs. Because CG talks in Pixels,
-    // curState->curTransform contains both the device scaling factor and any user scale applied to it causing the below D2D render target
-    // transform calculation to be way off on devices that have scaling. Workaround for now is to divide the verticalScalingFactor and
-    // height here with the device scaling factor so they get converted to DIPs and then apply the transform to the D2D render target.
-    // Though this workaround is sufficient to get text rendered on the user screen on all Windows form factor devices, this will not work
-    // when text is rendered on a PDF or printer.
-    static float hostDisplayScale = static_cast<float>([[WGDDisplayInformation getForCurrentView] resolutionScale] / 100.0f);
-    verticalScalingFactor /= hostDisplayScale;
-    height /= hostDisplayScale;
-    transform = CGAffineTransformTranslate(transform, 0, (height / verticalScalingFactor) - height);
-    // Perform anti-clockwise rotation required to match the reference platform.
-    imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, -transform.b, transform.c, transform.d, transform.tx, transform.ty));
+    // Only enter this branch when called by the user, not internal calls
+    // CoreText will only apply user created transformations, other internal transformations must be applied before drawing
+    // DirectWrite coordinate system matches all but CoreText, so the transformation needs to be inverted
+    if (invertedCoordinates) {
```

Comment:
DPI scaling is part of the handling the inverted coordinate system, so yes it belongs in here

---
In reply to: [83092992](https://github.com/Microsoft/WinObjC/pull/1132#discussion_r83092992) [](ancestors = 83092992)